util/src/main/java/com/google/cloud/hadoop/util/GcsJsonApiEvent.java

```diff
@@ -27,79 +29,77 @@
 import javax.annotation.Nonnegative;
 import javax.annotation.Nonnull;
 
-class GcsJsonApiEvent {
-  public enum EventType {
-    BACKOFF,
-    STARTED,
-    RETRYSKIPPED,
-    EXCEPTION,
-    RESPONSE,
-  }
-
+public class GcsJsonApiEvent implements IGcsJsonApiEvent {
   public static final String BACKOFFTIME = "BACKOFFTIME";
   public static final String RETRYCOUNT = "RETRYCOUNT";
   public static final String STATUS_CODE = "STATUS_CODE";
   public static final String DURATION = "DURATION";
+  public static final String REQUESTTYPE = "REQUESTTYPE";
   private final EventType eventType;
 
-  // Setting this to object so that we do not have to create the URL string.
+  // Having this as Object type so that we do not have to create the URL string.
   private final Object context;
-
   private final String method;
-
   private Map<String, Object> properties;
 
-  static GcsJsonApiEvent getResponseEvent(HttpResponse httpResponse, @Nonnegative long duration) {
-    GcsJsonApiEvent result = new GcsJsonApiEvent(httpResponse.getRequest(), EventType.RESPONSE, 2);
+  @VisibleForTesting
+  protected GcsJsonApiEvent(@Nonnull HttpRequest request, EventType eventType, Object context) {
+    this.eventType = eventType;
+    this.context = context;
+    this.method = request.getRequestMethod();
+  }
+
+  private GcsJsonApiEvent(HttpRequest request, EventType eventType, int capacity, Object context) {
+    this(request, eventType, context);
+    this.properties = new HashMap<>(capacity, 1);
+  }
+
+  @VisibleForTesting
+  public static GcsJsonApiEvent getResponseEvent(
+      HttpResponse httpResponse, @Nonnegative long duration, Object context) {
+    GcsJsonApiEvent result =
+        new GcsJsonApiEvent(httpResponse.getRequest(), EventType.RESPONSE, 2, context);
```

Comment:
should this be 3? also mind adding a comment like /* #properties */?

Reply:
I will add for the remaining asserts.

Reply:
Done. Thanks.